### ISSUE TITLE
use magics programmatically 

### ISSUE THREAD

**Issue by blazespinnaker** [ipython/ipython#13319]
Problem: 

IPython (and various notebook implementations) are self evidently a cornerstone of python programming itself. Creating libraries dependent on IPython is practiced widely across the data science industry for reasons I assume here are well known.

Unfortunately, there is lack of clear conventions for architectural cohesion when using extension based functionality.   Magics should be treated as first order APIs and not something outside standard python.  Basic concepts of cohesion wisely suggest keeping all code logic together. 

For example, I'm trying to use matplotlib inline set formats, and I don't see a clear and well documented way to do this programmatically.  

After a some digging, I found this a little hidden away: https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.interactiveshell.html
run_line_magic(magic_name, line, _stack_depth=1

That's great, but it only partially solves my problem.  I now need to call the matplotlib inline library programmatically to set the formats.  

For that, I was indirectly lead to some guidance by this for set_matplotlib_formats:  https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.set_matplotlib_formats
Which curiously tells me to do exactly what makes sense to me - call a python API.  Unfortunately, I'm stumped on the right way to get access to this library.

After some searching and using that deprecation notice as a hint, I found this: https://github.com/ipython/matplotlib-inline  

Unfortunately, it just mentions the magic and nothing about importing and using the APIs itself.   I dug into the github project and found the API call mentioned in the IPython documentation, but I'm still unsure of the right way to import and utilize this library.

I apologize if I am missing something obvious.  If that's the case, please just see this as a suggestion to highlight the documentation for API usage more enthusiastically.

**Comment by blazespinnaker** [comment]
I've also opened an issue with the matplotlib_inline github 

https://github.com/ipython/matplotlib-inline/issues/12

**Comment by ccordoba12** [comment]
You need to import the `get_ipython` function (which will be give you the IPython shell), then use it to call magics. You can see how we do that to set the Matplotlib backend in Spyder here:

https://github.com/spyder-ide/spyder-kernels/blob/45eacb1ce8058ec3e3394b6e6132255874400f87/spyder_kernels/console/kernel.py#L694-L746

**Comment by blazespinnaker** [comment]
Hmm, I think that's for matplotlib.   Matplotlib inline has "_enable_matplotlib_integration()" which appears to execute automatically on lib import.

**Comment by Carreau** [comment]
I don't think you should try to set inline matplotlib outside of IPython, and nowadays matplotlib should detect IPython on it's own, so you should almost never have to call `%matplotlib` magic.

To see what the "equivalent" python is for a magic, you can use `hist -t`, which will show you the post-transformed magics.

I agree with you that magics should be as much as possible treated as API, and I'd love to have more extensions points inside Python to do so. Unfortunately it looks like core python devs have always been agains anything macro-like, and for many magics we need to at least get the AST (timeit, for example), and even sometime the raw string.

It's also a bit weird as magics may also change the internal state of IPython (which is a bit the case with matplotlib magic).

All in all yes we'd really like to have a more decoupled ecosystem, it will just take time and efforts. 

I'll try to dig out the proper python way to call matplotlib inline without requiring IPython.
